### PR TITLE
Editable: Adding the link control

### PR DESCRIPTION
**Pull request by youknowriad**
This PR adds the link control to the formatting toolbar.
I had to change how the `Toolbar` works to allow advanced controls showing Modals and updating the state accordingly.

**Testing instructions**

 - Try adding links and navigating inside an Editable.

### CODE REVIEW

File: editor/modes/visual-editor/style.scss [hunk 1]
```diff
@@ -60,3 +60,22 @@
 .editor-visual-editor .editor-inserter {
 	margin: $item-spacing;
 }
+
+
+// joen make me pretty
```
Comment (youknowriad) [inline editor/modes/visual-editor/style.scss]: @jasmussen you shouldn't have showed me this tip :)

Comment (jasmussen) [inline editor/modes/visual-editor/style.scss]: Haha love it :D —  I'll take a look.

**Comment by jasmussen** [comment]
Nice work!

I'm curious on Ella's thoughts here, as the link feature of the current editor is _stellar_ in how it works. Whatever we could copy from that needs to come along for the ride. That includes the "select text and paste link" feature. There's also a bunch of little enhancements like the link dialog showing in context of the selected text, and easily dismissing with a click outside, as well as an easy way to unlink the link again. 

Here's a mockup:

screen shot 2017-04-25 at 14 01 47

To put it differently —  how can we port as much as possible of the current link behavior? Is that out of scope for this PR?

**Comment by jasmussen** [comment]
Here's a quick mockup of what I was thinking:

screen shot 2017-04-25 at 14 13 08

File: blocks/components/editable/index.js [hunk 2]
```diff
@@ -141,6 +141,14 @@ export default class Editable extends wp.element.Component {
 			return result;
 		}, {} );
 
+		// Link format
+		const link = parents.find( parent => parent.nodeName === 'A' );
```
Comment (aduth) [inline blocks/components/editable/index.js]: We're not polyfilling `Array#find`, this will error in IE11.

File: editor/components/toolbar/index.js [hunk 3]
```diff
@@ -9,27 +9,34 @@ import classNames from 'classnames';
 import './style.scss';
 import IconButton from 'components/icon-button';
 
-function Toolbar( { controls } ) {
+function Toolbar( { attributes, setAttributes, controls } ) {
```
Comment (aduth) [inline editor/components/toolbar/index.js]: My first reaction to these changes is that the line is becoming blurred between the concept of a general-purpose toolbar and a block-specific toolbar. We should decide if there's value in the former and whether we can or care to preserve this distinction.

File: editor/components/toolbar/index.js [hunk 4]
```diff
@@ -9,27 +9,34 @@ import classNames from 'classnames';
...
+						key={ index }
+						icon={ control.icon }
+						label={ control.title }
+						data-subscript={ control.subscript }
+						onClick={ ( event ) => {
+							event.stopPropagation();
+							control.onClick( attributes, setAttributes );
+						} }
+						className={ classNames( 'editor-toolbar__control', {
+							'is-active': control.isActive && control.isActive( attributes )
```
Comment (aduth) [inline editor/components/toolbar/index.js]: This will need to be rebased to account for changes in #500.

File: editor/modes/visual-editor/format-controls.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/**
```
Comment (aduth) [inline editor/modes/visual-editor/format-controls.js]: Should this file have been named `link-control.js`?

Comment (aduth) [inline editor/modes/visual-editor/format-controls.js]: Oh, I see now the `LinkControl` component is inlined in the file, but the default export is the array of formatting controls. Hmm...

File: editor/modes/visual-editor/format-controls.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+					onClick={ this.toggleLinkModal }
+					className={ classNames( 'editor-toolbar__control', {
+						'is-active': !! this.props.attributes.link
+					} ) }
+				/>
+				{ this.state.opened &&
+					<div className="editable-visual-editor__link-modal">
+						<form onSubmit={ this.submitLinkModal }>
+							<input type="url" value={ this.state.value } onChange={ this.updateLinkValue } />
+							<button>{ wp.i18n.__( 'Add Link' ) }</button>
```
Comment (aduth) [inline editor/modes/visual-editor/format-controls.js]: Trying to decide if there's value in the `<Button />` component (`editor/components/button`) being the preferred default usage so we can apply styles broadly to all buttons in the editor.

File: editor/modes/visual-editor/format-controls.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+				<IconButton
+					icon="admin-links"
+					label={ wp.i18n.__( 'Link' ) }
+					onClick={ this.toggleLinkModal }
+					className={ classNames( 'editor-toolbar__control', {
+						'is-active': !! this.props.attributes.link
+					} ) }
+				/>
+				{ this.state.opened &&
+					<div className="editable-visual-editor__link-modal">
```
Comment (aduth) [inline editor/modes/visual-editor/format-controls.js]: Can we apply this class to the `form` and drop the wrapper element?

**Review by aduth — COMMENTED** [review]
Empty file `editor/components/format-toolbar/index.js` should be removed.

File: editor/components/toolbar/index.js (hunk 3 above)
Comment (youknowriad) [inline editor/components/toolbar/index.js]: Well `attributes and `setAttributes` here don't mean "block attributes" and "update block attributes" but I see it as "toolbarState" "updateToolbarState" which is generic enough. Could be renamed but this would impact the `attributes` `setAttributes` used in the `isActive` and `onClick` of the block controls.

File: editor/modes/visual-editor/format-controls.js (hunk 6 above)
Comment (youknowriad) [inline editor/modes/visual-editor/format-controls.js]: Yes, this needs to be updated to match @jasmussen's comment

**Comment by youknowriad** [comment]
@jasmussen My question about your mockups is when do we show/hide the link "modal". Do we show it automatically when the link is focused?

**Comment by jasmussen** [comment]
> My question about your mockups is when do we show/hide the link "modal". Do we show it automatically when the link is focused?

I was thinking the behavior should be as 1:1 identical to the current wp-admin behavior as we can make it, to the point that if we could copy the whole thing over (I know that we can't) it would be great. 

link in wp admin

File: editor/components/toolbar/index.js (hunk 3 above)
Comment (aduth) [inline editor/components/toolbar/index.js]: >Well `attributes` and `setAttributes` here don't mean "block attributes" and "update block attributes" but I see it as "toolbarState" "updateToolbarState" which is generic enough.

Given that I had trouble making this distinction, I expect others will as well.

**Comment by youknowriad** [comment]
closing in favor of #524
